src/signTxOutput.c

```diff
@@ -0,0 +1,664 @@
+		        txHashBuilder,
+		        subctx->stateData.tokenGroup.policyId, MINTING_POLICY_ID_SIZE,
+		        subctx->numTokenAmounts
+		);
+		TRACE();
+	}
+
+	signTxOutput_handleTokenGroup_ui_runStep();
+}
+
+// ============================== TOKEN AMOUNT ==============================
+
+enum {
+	HANDLE_TOKEN_AMOUNT_STEP_DISPLAY_NAME = 800, // TODO
+	HANDLE_TOKEN_AMOUNT_STEP_DISPLAY_AMOUNT,
+	HANDLE_TOKEN_AMOUNT_STEP_RESPOND,
+	HANDLE_TOKEN_AMOUNT_STEP_INVALID,
+};
+
+static void signTxOutput_handleTokenAmount_ui_runStep()
+{
+	TRACE("UI step %d", subctx->ui_step);
+	ui_callback_fn_t* this_fn = signTxOutput_handleTokenAmount_ui_runStep;
+
+	UI_STEP_BEGIN(subctx->ui_step);
+
+	UI_STEP(HANDLE_TOKEN_AMOUNT_STEP_DISPLAY_NAME) {
+		token_amount_t* tokenAmount = &subctx->stateData.tokenAmount;
+		if (str_isTextPrintable(tokenAmount->assetName, tokenAmount->assetNameSize)) {
+			char name[ASSET_NAME_SIZE_MAX + 1];
+			ASSERT(tokenAmount->assetNameSize + 1 <= SIZEOF(name));
+			for (size_t i = 0; i < tokenAmount->assetNameSize; i++)
+				name[i] = tokenAmount->assetName[i];
+			name[tokenAmount->assetNameSize] = '\0';
+
+			ui_displayPaginatedText(
+			        "Token name",
+			        name,
+			        this_fn
+			);
+		} else {
+			ui_displayHexBufferScreen(
+			        "Token name",
+			        subctx->stateData.tokenAmount.assetName, subctx->stateData.tokenAmount.assetNameSize,
+			        this_fn
+			);
+		}
+	}
+	UI_STEP(HANDLE_TOKEN_AMOUNT_STEP_DISPLAY_AMOUNT) {
+		ui_displayUint64Screen(
+		        "Token amount",
+		        subctx->stateData.tokenAmount.amount,
+		        this_fn
+		);
+	}
+	UI_STEP(HANDLE_TOKEN_AMOUNT_STEP_RESPOND) {
+		respondSuccessEmptyMsg();
+
+		ASSERT(subctx->currentTokenAmount < subctx->numTokenAmounts);
+		subctx->currentTokenAmount++;
+
+		if (subctx->currentTokenAmount == subctx->numTokenAmounts) {
+			advanceState();
+		}
+	}
+	UI_STEP_END(HANDLE_TOKEN_AMOUNT_STEP_INVALID);
+}
+
+static void signTxOutput_handleTokenAmountAPDU(uint8_t* wireDataBuffer, size_t wireDataSize)
+{
+	{
+		// sanity checks
+		CHECK_STATE(STATE_OUTPUT_TOKEN_AMOUNT);
+
+		ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
+	}
+	{
+		token_amount_t* tokenAmount = &subctx->stateData.tokenAmount;
+
+		// parse data
+		TRACE_BUFFER(wireDataBuffer, wireDataSize);
+		read_view_t view = make_read_view(wireDataBuffer, wireDataBuffer + wireDataSize);
+
+		VALIDATE(view_remainingSize(&view) >= 4, ERR_INVALID_DATA);
+		tokenAmount->assetNameSize = parse_u4be(&view);
+		VALIDATE(tokenAmount->assetNameSize > 0, ERR_INVALID_DATA);
+		VALIDATE(tokenAmount->assetNameSize <= ASSET_NAME_SIZE_MAX, ERR_INVALID_DATA);
+
+		ASSERT(tokenAmount->assetNameSize <= SIZEOF(tokenAmount->assetName));
+		view_memmove(tokenAmount->assetName, &view, tokenAmount->assetNameSize);
+
+		VALIDATE(view_remainingSize(&view) == 8, ERR_INVALID_DATA);
+		tokenAmount->amount = parse_u8be(&view);
+		TRACE_UINT64(tokenAmount->amount);
+		// TODO validate something?
```

Comment:
did you have something specific in mind? If not I think we don't really have anything to validate as we don't have further context about the specific token